File: src/cli/tools/epic.go
// src/cli/tools/epic.go
package tools

import (
	"fmt"
	"io"
	"os"
	"path/filepath"

	"github.com/lucas/oraculo/src/db"
	"github.com/lucas/oraculo/src/output"
	"github.com/spf13/cobra"
)

// newEpicCmd returns the "epic" parent command with all subcommands.
func newEpicCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "epic",
		Short: "Manage epics",
	}
	cmd.AddCommand(
		newEpicInitCmd(),
		newEpicSaveCmd(),
		newEpicGetCmd(),
		newEpicListCmd(),
		newEpicUpdateCmd(),
		newEpicDeleteCmd(),
	)
	return cmd
}

func newEpicInitCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "init <name>",
		Short: "Create a new epic",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			name := args[0]
			description, _ := cmd.Flags().GetString("description")
			w := cmd.OutOrStdout()

			store := db.NewEpicStore(dbFromContext(cmd.Context()))
			_, created, err := store.Create(name, description)
			if err != nil {
				output.WriteError(w, err)
				return err
			}
			return output.WriteJSON(w, map[string]any{
				"name":    name,
				"path":    filepath.Join(".oraculo", "epics", name),
				"created": created,
			})
		},
	}
	cmd.Flags().String("description", "", "Epic description")
	return cmd
}

func newEpicSaveCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "save <name>",
		Short: "Save requirements markdown for an epic (reads stdin)",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			name := args[0]
			w := cmd.OutOrStdout()

			content, err := io.ReadAll(cmd.InOrStdin())
			if err != nil {
				output.WriteError(w, fmt.Errorf("read stdin: %w", err))
				return err
			}

			dir := filepath.Join(".oraculo", "epics", name)
			if err := os.MkdirAll(dir, 0o755); err != nil {
				output.WriteError(w, fmt.Errorf("create directory: %w", err))
				return err
			}

			path := filepath.Join(dir, "requirements.md")
			if err := os.WriteFile(path, content, 0o644); err != nil {
				output.WriteError(w, fmt.Errorf("write file: %w", err))
				return err
			}

			return output.WriteJSON(w, map[string]any{
				"name":  name,
				"path":  path,
				"saved": true,
			})
		},
	}
}

func newEpicGetCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "get <name>",
		Short: "Get requirements markdown for an epic",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			name := args[0]
			w := cmd.OutOrStdout()

			path := filepath.Join(".oraculo", "epics", name, "requirements.md")
			content, err := os.ReadFile(path)
			if err != nil {
				wrappedErr := fmt.Errorf("epic %q requirements: %w", name, err)
				output.WriteError(w, wrappedErr)
				return wrappedErr
			}

			_, err = w.Write(content)
			return err
		},
	}
}

func newEpicListCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "list",
		Short: "List all epics",
		Args:  cobra.NoArgs,
		RunE: func(cmd *cobra.Command, args []string) error {
			w := cmd.OutOrStdout()

			store := db.NewEpicStore(dbFromContext(cmd.Context()))
			epics, err := store.List()
			if err != nil {
				output.WriteError(w, err)
				return err
			}
			return output.WriteJSON(w, epics)
		},
	}
}

func newEpicUpdateCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "update <name>",
		Short: "Update an epic's description",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			name := args[0]
			description, _ := cmd.Flags().GetString("description")
			w := cmd.OutOrStdout()

			store := db.NewEpicStore(dbFromContext(cmd.Context()))
			epic, err := store.Update(name, description)
			if err != nil {
				output.WriteError(w, err)
				return err
			}
			return output.WriteJSON(w, epic)
		},
	}
	cmd.Flags().String("description", "", "New epic description")
	return cmd
}

func newEpicDeleteCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "delete <name>",
		Short: "Delete an epic",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			name := args[0]
			w := cmd.OutOrStdout()

			store := db.NewEpicStore(dbFromContext(cmd.Context()))
			if err := store.Delete(name); err != nil {
				output.WriteError(w, err)
				return err
			}
			return output.WriteJSON(w, map[string]any{
				"name":    name,
				"deleted": true,
			})
		},
	}
}

File: apps/orchestrator/src/cli/tools/epic.ts
import { Command } from "commander";
import { join } from "node:path";
import { mkdirSync } from "node:fs";
import { EpicStore } from "../../db/stores/epic-store.ts";
import { VersionStore } from "../../db/stores/version-store.ts";
import { ApprovalStore } from "../../db/stores/approval-store.ts";
import { writeJSON, writeError } from "../../utils/output.ts";
import { dbFromCmd } from "../context.ts";

export function epicCommands(): Command {
  const cmd = new Command("epic").description("Manage epics");

  cmd.addCommand(epicInitCmd());
  cmd.addCommand(epicSaveCmd());
  cmd.addCommand(epicGetCmd());
  cmd.addCommand(epicListCmd());
  cmd.addCommand(epicUpdateCmd());
  cmd.addCommand(epicDeleteCmd());
  cmd.addCommand(epicVersionCmd());
  cmd.addCommand(epicVersionsCmd());

  return cmd;
}

function epicInitCmd(): Command {
  return new Command("init")
    .description("Create a new epic")
    .argument("<name>", "Epic name")
    .option("--description <desc>", "Epic description", "")
    .action((name: string, opts: any, cmd: Command) => {
      try {
        const db = dbFromCmd(cmd);
        const store = new EpicStore(db);
        store.create(name, opts.description);
        writeJSON({
          name,
          path: join(".oraculo", "epics", name),
          created: true,
        });
      } catch (err) {
        writeError(err);
      }
    });
}

function epicSaveCmd(): Command {
  return new Command("save")
    .description("Save requirements markdown for an epic (reads stdin)")
    .argument("<name>", "Epic name")
    .action(async (name: string) => {
      try {
        const content = await Bun.stdin.text();
        const dir = join(".oraculo", "epics", name);
        mkdirSync(dir, { recursive: true });
        const path = join(dir, "requirements.md");
        await Bun.write(path, content);
        writeJSON({ name, path, saved: true });
      } catch (err) {
        writeError(err);
      }
    });
}

function epicGetCmd(): Command {
  return new Command("get")
    .description("Get requirements markdown for an epic")
    .argument("<name>", "Epic name")
    .action(async (name: string) => {
      try {
        const path = join(".oraculo", "epics", name, "requirements.md");
        const file = Bun.file(path);
        if (!(await file.exists())) {
          writeError(new Error(`epic "${name}" requirements: file not found`));
          return;
        }
        const content = await file.text();
        process.stdout.write(content);
      } catch (err) {
        writeError(err);
      }
    });
}

function epicListCmd(): Command {
  return new Command("list")
    .description("List all epics")
    .action((opts: any, cmd: Command) => {
      try {
        const db = dbFromCmd(cmd);
        const store = new EpicStore(db);
        writeJSON(store.list());
      } catch (err) {
        writeError(err);
      }
    });
}

function epicUpdateCmd(): Command {
  return new Command("update")
    .description("Update an epic's description")
    .argument("<name>", "Epic name")
    .option("--description <desc>", "New epic description", "")
    .action((name: string, opts: any, cmd: Command) => {
      try {
        const db = dbFromCmd(cmd);
        const store = new EpicStore(db);
        const epic = store.update(name, opts.description);
        writeJSON(epic);
      } catch (err) {
        writeError(err);
      }
    });
}

function epicDeleteCmd(): Command {
  return new Command("delete")
    .description("Delete an epic")
    .argument("<name>", "Epic name")
    .action((name: string, opts: any, cmd: Command) => {
      try {
        const db = dbFromCmd(cmd);
        const store = new EpicStore(db);
        store.delete(name);
        writeJSON({ name, deleted: true });
      } catch (err) {
        writeError(err);
      }
    });
}

function epicVersionCmd(): Command {
  return new Command("version")
    .description("Create a new version of epic requirements (reads stdin)")
    .argument("<name>", "Epic name")
    .action(async (name: string, opts: any, cmd: Command) => {
      try {
        const db = dbFromCmd(cmd);
        const epicStore = new EpicStore(db);
        const epic = epicStore.getByName(name);

        const content = await Bun.stdin.text();

        const versionStore = new VersionStore(db);
        const version = versionStore.CreateEpicVersion(epic.id, content);

        const approvalStore = new ApprovalStore(db);
        const approval = approvalStore.Create("design", epic.id, null, content);

        writeJSON({
          version_id: version.id,
          approval_id: approval.id,
        });
      } catch (err) {
        writeError(err);
      }
    });
}

function epicVersionsCmd(): Command {
  return new Command("versions")
    .description("List all versions of an epic")
    .argument("<name>", "Epic name")
    .action((name: string, opts: any, cmd: Command) => {
      try {
        const db = dbFromCmd(cmd);
        const epicStore = new EpicStore(db);
        const epic = epicStore.getByName(name);

        const versionStore = new VersionStore(db);
        writeJSON(versionStore.ListEpicVersions(epic.id));
      } catch (err) {
        writeError(err);
      }
    });
}
